core/src/code.rs

```diff
@@ -97,19 +105,47 @@ impl Code {
                 .ok_or(CodeError::MemoryEntryNotFound)?,
         );
 
-        let gas_rules = get_gas_rules(&module);
-        let instrumented_module = wasm_instrument::gas_metering::inject(module, &gas_rules, "env")
-            .map_err(|_| CodeError::GasInjection)?;
-
-        let instrumented = wasm_instrument::parity_wasm::elements::serialize(instrumented_module)
-            .map_err(|_| CodeError::Encode)?;
+        let mut exports = BTreeSet::<DispatchKind>::new();
+
+        for entry in module
+            .export_section()
+            .ok_or(CodeError::ExportSectionNotFound)?
+            .entries()
+            .iter()
+        {
+            if let Internal::Function(_) = entry.internal() {
+                if entry.field() == DispatchKind::Init.into_entry() {
+                    exports.insert(DispatchKind::Init);
+                } else if entry.field() == DispatchKind::Handle.into_entry() {
+                    exports.insert(DispatchKind::Handle);
+                } else if entry.field() == DispatchKind::Reply.into_entry() {
+                    exports.insert(DispatchKind::Reply);
+                } else {
+                    return Err(CodeError::NonGearExportFnFound);
+                }
+            }
+        }
 
-        Ok(Self {
-            code: instrumented,
-            raw_code,
-            static_pages,
-            instruction_weights_version: version,
-        })
+        if exports.contains(&DispatchKind::Init) || exports.contains(&DispatchKind::Handle) {
+            let gas_rules = get_gas_rules(&module);
+            let instrumented_module =
+                wasm_instrument::gas_metering::inject(module, &gas_rules, "env")
+                    .map_err(|_| CodeError::GasInjection)?;
+
+            let instrumented =
+                wasm_instrument::parity_wasm::elements::serialize(instrumented_module)
+                    .map_err(|_| CodeError::Encode)?;
+
+            Ok(Self {
+                code: instrumented,
+                raw_code,
+                exports,
+                static_pages,
+                instruction_weights_version: version,
+            })
+        } else {
+            Err(CodeError::RequiredExportFnNotFound)
+        }
```

Comment:
If I'm not mistaken then this part of code could be extracted to the new standalone method and called here and in `new_raw` accrodingly.